concrete/src/Http/Middleware/ThumbnailMiddleware.php

```diff
@@ -60,20 +63,23 @@ public function process(Request $request, DelegateInterface $frame)
         $response = $frame->next($request);
 
         if ($this->app->isInstalled()) {
-            if ($response->getStatusCode() == 200) {
-                /* @var Connection $database */
-                try {
-                    $database = $this->getConnection();
-                } catch (\InvalidArgumentException $e) {
-                    // Don't die here if there's no available database connection
-                    $database = null;
-                }
-
-                if ($database) {
+            $responseStatusCode = (int) $response->getStatusCode();
+            if ($responseStatusCode === 200 || $responseStatusCode === 404) {
+                $database = $this->tryGetConnection();
+                if ($database !== null) {
                     try {
-                        $paths = $database->executeQuery('SELECT * FROM FileImageThumbnailPaths WHERE isBuilt=0 LIMIT 5');
-                        if ($paths->rowCount()) {
-                            $this->generateThumbnails($paths, $database);
+                        if ($responseStatusCode === 404) {
+                            $searchThumbnailPath = $request->getRequestUri();
+                        } else {
+                            $searchThumbnailPath = null;
+                        }
+                        $paths = $this->getThumbnailPathsToGenerate($database, $searchThumbnailPath);
```

Comment:
I would do `iterator_to_array($this->getThumbnai....);` so that you can loop over this result set more than the one time.